### PR TITLE
iOS failback support

### PR DESCRIPTION
**Pull request by zsalab**
This pull request add iOS failback support to this project. The modified WebcamJS detect the iOS devices (no flash and no HTML5 userMedia support) and render an input field as a failback with special properties, so no preview but on the capture event the iOS offer to use the camera to take a picture and this step provide the preview to the user.

### CODE REVIEW

**Comment by erickdeoliveiraleal** [comment]
fixes #2050

**Comment by sunnykgupta** [comment]
Any updates on this?

**Comment by zsalab** [comment]
pease review it and marge if possible

**Comment by jhuckaby** [comment]
Thank you for this PR, @zsalab.  Sorry it took me so long to get around to reviewing it.  It seems to work, but I had some trouble due to UX issues, and I have some questions about the code.

First, it took me a long time to realize that I had to actually call `Webcam.snap()` to initiate the iOS dialog to use my phone's camera.  I was clicking and double-clicking all over the preview area and nothing was happening.  Could we add a `ontouchstart` handler on the main preview DIV, so touching it calls `input.click()`?  I think that tiny change would make this worlds easier for people to figure out how to use.

Can we change the "*No preview available*" text?  If we implement the `ontouchstart` thing, then how about setting it to "*Click here to open camera.*" or words to that effect?

What on earth is `iOS_defImg`?  it seems to be some kind of Base64 encoded data blob, but it isn't actually used anywhere in the code.  You assign it to a local variable, then that is never used.  Was this a previous attempt at something?  Can we remove it?

What's with all the `ondblclick` handlers?  I can't figure out what you are doing in the code with these, and this is what led me to believe I could double-click the preview area to open the camera.

Thanks again.  If we can make these changes I will merge the PR.

\- Joe

**Comment by zsalab** [comment]
Thank you for your review, you right with many things, unfortunately I'm not a JS developer and I also left some unused code there.
I updated my code based on your requests, please review it.

1., In the iOS simulator the "touchstart" event not works, but looks the click works well so I chose this one. Originally the plan was to show the picture selector just if snap method called to avoid the double click (once for select a picture and one for create the "snapshot"). Of course it sounds crazy but no callback information not exist if the user click on the "Click here to open camera" placeholder, because this come trough the snap method (by the time an external code call it). I changed the code and now works this way:
1., if no image selected the snap method offer the selection and create the "snapshot"
2., if somebody click to the "Click here to open camera" placeholder it can select the image but because no callback information the image just loaded to the placeholder area and the "real" "snapshot" created if the snap method called (and we get the callback information).

2., I've replaced the text

3., Fixed. Sorry I had many ideas to solve the problem and one used that what I've throw away for a while, now I removed the fake image data too.

4., It was a hack for pass the event handler (I'm not a JS developer so that time I didn't have better idea to solve the problem), I guess now the code more clean

Please let me know if you have more questions or other problems with the code. Any help welcome, especially I'm not JS expert.

**Comment by jhuckaby** [comment]
Okay, first of all, thank you @zsalab for all your work on this PR.  I am really impressed how you are using an image object, and passing that to snap().  Very clever solution to the problem.  That being said, I am not quite ready to accept this PR into the main branch.  Here is why:

(1) The implementation is still not quite right.  All my iPhone images are coming in rotated sideways, for example, and cropped horribly.  There is a also huge black box under them.  I suspect this is because the iOS camera doesn't adhere to the constraints, and is just providing us an image at whatever orientation the phone is at, and at native camera resolution.  WebcamJS is expecting an image at the proper aspect ratio and dimensions, so it is just falling over with iOS camera captures.  Screenshot below.

(2) The iOS capture doesn't support many of the WebcamJS features like scaling.  For example, the developer may want a large image, but only show a small preview.  The iOS implementation doesn't support the scaling features.

(3) This implementation is very iOS specific, and relies on a user agent sniff.  What about Android, and other mobile platforms?  I think this should ultimately be platform-agnostic, and just kick in if both getUserMedia and Flash both fail, *and* the developer has enabled a flag (i.e. `enable_mobile` or some such thing).

(4) I want to do more with the mobile UI, so it has a border, background, and perhaps a camera icon inside of it.  This should also be user-customizable via CSS, so the developer can blend it in with their site's look & feel.

(5) Using the "click" event adds a huge delay (latency) on mobile, because it is waiting for gestures like double-click, etc.  That is why I recommended `touchstart`, as that is instantaneous.  I am not sure why this is failing on your iOS simulator.

(6) I am thinking maybe there shouldn't be a local image preview at all.  Meaning, capturing an image just triggers snap() right away.  The camera on the mobile device already allows the user to preview and choose to keep or discard the image in the native OS dialog.  At that point, when we get the change event, I think we should just snap right away, and not even show a local preview.

Here is what I am getting using my iPhone 7, taking pics in portrait mode:

img_1110

As you can see the captured image is rotated 90 degrees, and has a huge black box on the bottom.  Yuck!

In this case the WebcamJS page wants a 320x240 image, but that "constraint" is not known or supported by the iOS camera, and it instead just delivers a image at whatever resolution and aspect ratio the phone is in.  So we're going to need some more special-case code to auto-scale and auto-crop the image to meet the constraints, emulating what getUserMedia and Flash do natively.

Anyway, in summary I am very impressed by this PR, and I am amazed you got it to work at all, but this is not ready to merge into the main branch.  I am terrified at the sheer number of user complaints and issues I will get flooded with.

I've been slowly working on WebcamJS 2.0, which has support for mobile using a technique similar to yours.  It's not at all ready for release, but I am thinking maybe I should just keep working on that branch.

In the meantime, I am more than happy to add something to the main README, pointing people at your branch, if they need iOS support now, and cannot wait for v2.

**Comment by zsalab** [comment]
Thank you for the review

1., I think I can handle this with auto crop and a hidden full size image.

2., This too

3., Of course it is iOS specific, how far I know Android supports userMedia but iOS not. I don't have Android device and our development not targeting any kind of Android devices. Refuse the pull request based on this won't be fair. Please let me know if it a blocker for the merge.

4., I think I can manage this

5., I gonna try on my iPhone, but I think it should not block the merge (but of course it is your code base)

6., You are absolutely right but I need a code right now which can easily merge with your code base, if I change the callback handling (e.g. pass the callback information at init), I can solve this and break the API compatibility. Right now I don't want to break the API and I want to keep this patch clean as possible to easily follow the main codebase (that's why you can see interesting solution, I didn't want to modify the main code). Unfortunately not possible to control the camera from the browser on iOS at all (which camera, resolution, size, orientation, etc... - nothing). 

I continue on this if you can accept to merge with bullet points 1, 2 and 4 fixed/improved. But you must know, based on the iOS limitations not possible to add a perfect solution. I think if the documentation is properly explain the limitations on iOS and show that because of the iOS limitations should not be a problem, but of course it is your decision.

If no chance to merge with the limitations, I'm stop working on this (I don't want to waste my time)

**Comment by zsalab** [comment]
Please review it again.

1., fixed

2., fixed

3., my focus is iOS because right now is totally not supported (Android should work with HTML5 userMedia)

4., all of the visible elements got CSS class, also the text of the preview can pass as a parameter, any kind of customization is supported.

5., touchstart does not work for me, maybe who better on this can change later, with the click event is works well

6., because on iOS the camera is not accessible at all, maybe the preview can be used for positioning and cropping the image. Right now it uses auto crop (center of the image).
It also possible development to pre define the snap callback trough class parameter and in this case it would be possible to create the final image trough the preview placeholder.

Please merge if it is possible, I think this version can be a good beta version of the iOS support.

**Comment by jhuckaby** [comment]
Oh wow, you did it!  You work very fast.  I am quite impressed.  I thought these features would take weeks.  I need some time to review this, and will check it out this weekend.  Thanks man!

**Comment by jhuckaby** [comment]
Hey @zsalab, so I feel really bad.  You've put in all this amazing work, and are doing things I didn't even think possible with WebcamJS, but it still isn't working well enough to accept into the master branch.  When I take a picture with my iPhone using your latest code, it gets smashed (distorted) to fit the size of the container.  This is a very poor user experience.  What should happen is a "cover" resize (scale down to cover the container, then auto-crop off the bleed).  This is what getUserMedia and Flash do natively, and what I am working on with WebcamJS 2.0.

Also, there are still quite a few UI / UX quirks here.  For example, we're still showing a preview image on iOS, after having already been shown a preview in the native camera capture UI.  This makes the user jump through too many hoops to get their image submitted.  I think it should auto-snap and not even show a preview in the page.  Also, when you are showing the local preview, the "Click here to open camera" text is still superimposed on top, barely legible.  It just looks and feels wrong to me.

@zsalab, I don't want you to waste anymore of your time with this.  I feel terrible, because this is incredible work, and frankly you should be paid for it.  But I just can't accept this as is.  It needs a lot of cleanup, it needs to handle edge cases, the code needs to be commented, and I am also worried that we're "borrowing" 3rd party code from Exif.js and possibly other places.

My last point I have to make about this is, when I merge a PR, I have to fully understand the code.  This is because I have to support it going forward.  But I don't really understand this code, or what it is doing.

At this point I would say, please don't spend anymore time on this.  I cannot merge this PR, but I am extremely thankful for your work.  Also, please, please let me pay you for the work you have put in so far.  Do you have a PayPal account?  I would really like to give you something for your efforts, because this is truly awesome work, and you should be rewarded.  Please PM me your PayPal.

Again, thank you so much, and I am terribly sorry I cannot merge this at this time.  If you want, I would be more than happy to link to your fork of WebcamJS from the main documentation.

Please have a wonderful weekend, and I am sorry.

\- Joe

**Comment by zsalab** [comment]
Probably you have a good reason why you not willing to accept this.

1., I tried on my iPhone and also on iOS simulator devices, I don't know what you mention about the poor quality. What this code do that all of the code do if scaling a picture to a specific size (scale and crop)... of course maybe a bug, but probably I not was careful to write the resize+crop code.... probably it 15 minutes to fix it.
2., UI/UX problem what you mention is not a big deal, I can easily add callback parameter to the class and take the snapshot immediately. Hiding the text also not a big deal. So we talking about 15 minutes again
3., your code also does not contains to much comments, but anyway I could add comments in 10 minutes
4., NO CODE "BARROWED" from anywhere, I got the algorithm what is defined in open document by JEIDA, but I not speak Japanese and this was the fastest way, but you can not found this code in exif-js, I mention them in the code because they save a lot of time for me to get know how can I extract the orientation information from the file.
5., I don't care about the money, I used your code (looks unfortunately) trough primefaces and I needed iOS support I added and I want to give back to the open source "community".

This is your shame!

By the way, I will fix and improved my codebase on the next week based on your feedback and thoughts, but I'm not willing to argue with you, I not gonna ask any time to merge and I will ask primefaces to revoke my code from them codebase. I will have my custom component and nobody will enjoy! Great open source world.

**Comment by zsalab** [comment]
iOS demo for who interested (no "take snapshot button" - e.g works only on iOS)
https://zettasystem.com/webcamjs/demos/ios-large.html
iOS and other browsers (HTML5 video or Flash)
https://zettasystem.com/webcamjs/demos/large.html

**Comment by jhuckaby** [comment]
Looks great, thank you very much for your contribution.

**Comment by zsalab** [comment]
Thank you so much to merge. I know you want to focus on 2.0 and I understand this.
I would like to offer you my contribution about the merged iOS support (for free of charge), so any time when you get an issue with this code, you can assign to me and I gonna fix soon as possible. You can focus on the 2.0

**Comment by jhuckaby** [comment]
Thank you, @zsalab!  I really appreciate that, and thank you for all your hard work on this.

**Comment by sivajik34** [comment]
@zsalab webcam not working in Ipad chrome.
Ipad safari ,Mac chrome,safari its working fine.can you look into this?

**Comment by zsalab** [comment]
Sorry I do not have iPad and it isn't simple to install a Chrome to the Simulator. It would be nice to have least some errors on it (from a debugger).... but anyway the iOS part of the WebcamJS is developed for Safari, I never tested (and planned to test) in other browsers on iOS 
So if you be able to provide some error messages maybe I can do something, but please add a new issue about this and mention me in the description.... and I will try to do the best
